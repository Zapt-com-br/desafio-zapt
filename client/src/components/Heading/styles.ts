import styled, { css } from 'styled-components'
import { HeadingProps } from '.'

export const Wrapper = styled.h1.attrs<HeadingProps>(({ level }) => ({
  as: `h${level}`
}))<HeadingProps>`
  ${({ theme, color, size, fontWeight, lineHeight }) => css`
    font-size: ${theme.font.sizes[size!]};
    color: ${color};
    font-weight: ${fontWeight};
    line-height: ${lineHeight};
  `}
`
