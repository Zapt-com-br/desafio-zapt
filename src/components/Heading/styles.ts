import styled, { css, DefaultTheme } from 'styled-components';
import media from 'styled-media-query';

import { HeadingProps } from '.';

export const wrapperModifiers = {
  small: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.medium};
  `,

  medium: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.xlarge};

    ${media.greaterThan('medium')`
      font-size: ${theme.font.sizes.xxlarge};
    `}
  `,

  huge: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.huge};
  `,
};

export const Wrapper = styled.h2<HeadingProps>`
  ${({ theme, size, weight }) => css`
    color: ${theme.colors.black};
    font-weight: ${theme.font[weight!]};

    ${!!size && wrapperModifiers[size](theme)}
  `}
`;
