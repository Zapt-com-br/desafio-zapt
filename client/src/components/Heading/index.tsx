import theme from 'styles/theme'
import * as S from './styles'

export type HeadingProps = {
  color?: string
  size?: keyof typeof theme.font.sizes
  fontWeight?: 300 | 400 | 500
  lineHeight?: number | string
  level?: 1 | 2 | 3 | 4 | 5 | 6
  children: React.ReactNode
}

const Heading = ({
  color = '#000000',
  size = 'normal',
  fontWeight = 500,
  lineHeight = 1.5,
  level = 1,
  children
}: HeadingProps) => (
  <S.Wrapper
    level={level}
    color={color}
    size={size}
    fontWeight={fontWeight}
    lineHeight={lineHeight}
  >
    {children}
  </S.Wrapper>
)

export default Heading
