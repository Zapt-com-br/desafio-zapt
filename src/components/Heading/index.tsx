import * as S from './styles';

export type HeadingProps = {
  children: React.ReactNode;
  size?: 'small' | 'medium' | 'huge';
  weight?: 'light' | 'normal' | 'medium';
};

const Heading = ({
  children,
  size = 'medium',
  weight = 'normal',
}: HeadingProps) => (
  <S.Wrapper size={size} weight={weight}>
    {children}
  </S.Wrapper>
);

export default Heading;
