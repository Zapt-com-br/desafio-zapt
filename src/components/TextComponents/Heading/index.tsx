import React from 'react';
import * as Styled from './styles';

export type HeadingProps = {
  text: string;
  size?: 'small' | 'large';
  bold?: '300' | '400' | '500';
};

const Heading = ({ text, size = 'small', bold = '400' }: HeadingProps) => (
  <Styled.Heading bold={bold} size={size}>
    {text}
  </Styled.Heading>
);

export default Heading;
