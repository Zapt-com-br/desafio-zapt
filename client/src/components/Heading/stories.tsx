import { Story, Meta } from '@storybook/react'
import Heading, { HeadingProps } from '.'

export default {
  title: 'Heading',
  component: Heading,
  argTypes: {
    children: {
      type: 'string'
    },
    size: {
      options: [
        'xsmall',
        'small',
        'medium',
        'normal',
        'large',
        'xlarge',
        'huge'
      ],
      control: { type: 'select' },
      defaultValue: 'normal'
    },
    level: {
      options: [1, 2, 3, 4, 5, 6],
      control: { type: 'select' },
      defaultValue: 1
    },
    color: {
      control: 'color',
      defaultValue: '#000000'
    },
    fontWeight: {
      name: 'Font Weight',
      options: [300, 400, 500],
      control: { type: 'select' },
      defaultValue: 500
    },
    lineHeight: {
      name: 'Line Height',
      type: 'number',
      defaultValue: 1.5
    }
  }
} as Meta

export const Default: Story<HeadingProps> = (args) => <Heading {...args} />

Default.args = {
  children: 'Bem Vindo Alan Gabriel'
}
