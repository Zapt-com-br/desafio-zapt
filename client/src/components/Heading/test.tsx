import { screen } from '@testing-library/react'
import renderWithTheme from 'utils/tests/test-util'

import Heading from '.'

describe('<Heading />', () => {
  it('should render heading as h1 by default', () => {
    const { container } = renderWithTheme(<Heading>Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 1 })
    ).toBeInTheDocument()

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 1 })
    ).toHaveStyle({
      color: '#000000',
      'font-size': '1.8rem',
      'font-weight': 500
    })

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render heading as h2 when level 2 is passed', () => {
    renderWithTheme(<Heading level={2}>Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 2 })
    ).toBeInTheDocument()
  })

  it('should render heading as h3 when level 3 is passed', () => {
    renderWithTheme(<Heading level={3}>Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 3 })
    ).toBeInTheDocument()
  })

  it('should render heading as h4 when level 4 is passed', () => {
    renderWithTheme(<Heading level={4}>Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 4 })
    ).toBeInTheDocument()
  })

  it('should render heading as h5 when level 5 is passed', () => {
    renderWithTheme(<Heading level={5}>Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 5 })
    ).toBeInTheDocument()
  })

  it('should render heading as h5 when level 5 is passed', () => {
    renderWithTheme(<Heading level={5}>Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 5 })
    ).toBeInTheDocument()
  })

  it('should render heading as h6 when level 6 is passed', () => {
    renderWithTheme(<Heading level={6}>Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 6 })
    ).toBeInTheDocument()
  })

  it('should render heading with the color "white" when color white is passed', () => {
    renderWithTheme(<Heading color="white">Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 1 })
    ).toHaveStyle({
      color: 'white'
    })
  })

  it('should render heading with size "4.0rem" when size huge is passed', () => {
    renderWithTheme(<Heading size="huge">Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 1 })
    ).toHaveStyle({
      'font-size': '4.0rem'
    })
  })

  it('should render heading with fontWeight "300" when passed', () => {
    renderWithTheme(<Heading fontWeight={300}>Bem vindo</Heading>)

    expect(
      screen.getByRole('heading', { name: /bem vindo/i, level: 1 })
    ).toHaveStyle({
      'font-weight': 300
    })
  })
})
