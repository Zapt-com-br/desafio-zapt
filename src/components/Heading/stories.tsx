import { Story, Meta } from '@storybook/react/types-6-0';
import Heading, { HeadingProps } from '.';

export default {
  title: 'Heading',
  component: Heading,
  argTypes: {
    children: {
      type: 'string',
    },
    size: {
      type: '',
    },
  },
} as Meta;

export const Default: Story<HeadingProps> = ({ children }) => (
  <>
    <Heading size="huge">Huge - {children}</Heading>
    <Heading>Medium - {children}</Heading>
    <Heading size="small">Small - {children}</Heading>
  </>
);

Default.args = {
  children: 'Heading',
};
