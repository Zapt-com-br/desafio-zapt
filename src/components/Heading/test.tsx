import { screen } from "@testing-library/react";
import { renderWithTheme } from "utils/tests/helpers";

import Heading from ".";

describe("<Heading />", () => {
  it("should render a white heading by default", () => {
    renderWithTheme(<Heading>Zapt</Heading>);
    expect(screen.getByRole("heading", { name: /Zapt/i })).toHaveStyle({
      color: "#ffffff",
    });
  });

  it("should render a black heading when color is passed", () => {
    renderWithTheme(<Heading color="black">Zapt</Heading>);
    expect(screen.getByRole("heading", { name: /Zapt/i })).toHaveStyle({
      color: "#000000",
    });
  });

  it("should render a gray heading when gray color is passed", () => {
    renderWithTheme(<Heading color="gray">Zapt</Heading>);
    expect(screen.getByRole("heading", { name: /Zapt/i })).toHaveStyle({
      color: "#7B7B7B",
    });
  });
});
