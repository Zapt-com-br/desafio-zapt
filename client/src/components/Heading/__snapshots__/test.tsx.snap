// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Heading /> should render heading as h1 by default 1`] = `
.c0 {
  font-size: 1.8rem;
  color: #000000;
  font-weight: 500;
}

@media (max-width:550px) {
  .c0 {
    font-size: calc(1.8rem / 1.3);
    line-height: 1.3;
  }
}

<h1
  class="c0"
  color="#000000"
  font-weight="500"
>
  Bem vindo
</h1>
`
